subsystem/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+
+//! General framework for working with subsystems
+
+use std::future::Future;
+use tokio::sync::{broadcast, mpsc, oneshot};
+use tokio::task;
+
+use logging::log;
+
+/// Top-level subsystem manager.
+///
+/// An application is composed of a number of long-lived subsystems. The [Manager] type starts
+/// and manages the life cycle of the subsystems. Whenever a subsystem exits, all other subsystems
+/// are requested to terminate and the manager is shut down.
+pub struct Manager {
+    // Manager name
+    name: &'static str,
+
+    // Order all subsystems to shutdown
+    shutdown_request_tx: broadcast::Sender<()>,
+    shutdown_request_rx: broadcast::Receiver<()>,
+
+    // Signal a subsystem is shutting down. This is taken as an instruction for all subsystems to
+    // shut down. Shutdown completion is detected by all senders having closed this channel.
+    shutting_down_tx: mpsc::Sender<()>,
+    shutting_down_rx: mpsc::Receiver<()>,
+}
+
+impl Manager {
+    /// Initialise a new subsystem manager.
+    pub fn new(name: &'static str) -> Self {
+        log::info!("Initialising subsystem manager {}", name);
+
+        let (shutdown_request_tx, shutdown_request_rx) = broadcast::channel(1);
+        let (shutting_down_tx, shutting_down_rx) = mpsc::channel(1);
+
+        Self {
+            name,
+            shutdown_request_tx,
+            shutdown_request_rx,
+            shutting_down_tx,
+            shutting_down_rx,
+        }
+    }
+
+    /// Build a new subsystem
+    ///
+    /// This gives more control over starting subsystems than [Manager::start] or
+    /// [Manager::start_passive]. Use if extra parameter tuning is required.
+    pub fn builder(&self) -> Builder<'_> {
+        Builder::new(self)
+    }
+
+    /// Start a subsystem. See [Builder::start].
+    pub fn start<T: 'static + Send, F: 'static + Send + Future<Output = ()>>(
+        &self,
+        name: &'static str,
+        subsystem: impl 'static + Send + FnOnce(CallRequest<T>, ShutdownRequest) -> F,
+    ) -> Subsystem<T> {
+        self.builder().with_name(name).start(subsystem)
+    }
+
+    /// Start a passive subsystem. See [Builder::start_passive].
+    pub fn start_passive<T: 'static + Send>(&self, name: &'static str, obj: T) -> Subsystem<T> {
+        self.builder().with_name(name).start_passive(obj)
+    }
+
+    async fn wait_for_subsystems_to_shut_down(mut shutting_down_rx: mpsc::Receiver<()>) {
+        // Wait for the subsystems to go down, signalled by closing the shutting_down channel.
+        while let Some(()) = shutting_down_rx.recv().await {}
+    }
+
+    /// Run the application main task.
+    ///
+    /// Completes when all the subsystems are fully shut down.
+    pub async fn main(mut self) {
+        // The main task just waits for the shutdown signal and coordinates the cleanup at the end.
+        // All other functionality is performed by subsystems.
+
+        // Install the Ctrl-C handler as a subsystem.
+        #[cfg(not(loom))]
```

Comment:
Why do we need this annotation?

Reply:
Loom does not emulate signals (or any other IO for that matter), so does not implement the `Ctrl-C` handling. Therefore, if you want to run something under loom, it can only spawn tasks that communicate through the synchronisation primitives provided by loom. No keyboard input, no network, no file IO, no signals etc.

Reply:
Now thinking about it, instead of disabling the subsystem under loom, we could make it an empty subsystem that immediately terminates. Then loom would try to schedule it at every possible point, effectively emulating trying to hit `Ctrl-C` at every point. Will give it a go and see what happens. Just note that loom does not currently handle this crate properly because the channels are not swapped out for the loom version when compiled for loom.

Reply:
Not sure I quite got it... It seems to me like the meaning of the annotation is that when running loom tests, the call to `start` should be skipped?
What would be the consequence of omitting the annotation?

Reply:
> Not sure I quite got it... It seems to me like the meaning of the annotation is that when running loom tests, the call to `start` should be skipped? What would be the consequence of omitting the annotation?

You would not be able to compile this with loom

Reply:
I removed the annotation and clippy and cargo build still succeed...

Reply:
Never mind, the loom tests don't compile.